client/src/document.js

```diff
@@ -100,69 +100,93 @@ function RenderSideBar({ doc }) {
     return null;
   }
   return doc.related_content.map(node => (
-    <SidebarLeaf
-      key={node.title}
-      depth={0}
-      title={node.title}
-      content={node.content || []}
-    />
+    <SidebarLeaf key={node.title} parent={node} />
   ));
 }
 
-function SidebarLeaf({ title, content }) {
-  const titleNode = <h3>{title}</h3>;
-  return (
-    <div>
-      {titleNode}
-      <ul>
-        {content.map(node => {
-          if (node.content) {
-            return (
-              <li key={node.title}>
-                <SidebarLeaflets node={node} />
-              </li>
-            );
-          } else {
-            return (
-              <li key={node.uri}>
-                <Link to={node.uri}>{node.title}</Link>
-              </li>
-            );
-          }
-        })}
-      </ul>
-    </div>
-  );
-}
-
-function SidebarLeaflets({ node }) {
+function SidebarLeaf({ parent }) {
+  const titleNode = <h3>{parent.title}</h3>;
   return (
     <Location>
       {({ location }) => {
-        let hasActiveChild = false;
-        const listItems = node.content.map(childNode => {
-          const isActive = childNode.uri === location.pathname;
-          if (isActive && !hasActiveChild) {
-            hasActiveChild = true;
+        // walk through the tree, annotating its nodes:
+        // * `isActive` if a node contains the link to the current page
+        // * `open` if a node is on the path to the `isActive` node
+        function setOpenNodes(node, pathname) {
```

Comment:
Stylistically, I don't like that it's called `setOpenNodes` and then it's used as a getter function, ie. you care about its return value. 

Creating functions on-the-fly isn't great for performance but the effect is unnoticeable at this scale. But I think it's worth keeping fast-by-design in mind if this is ever going to scale. 
The only variable here is the `location.pathname` which gets copied even though it's available as a closure. 

The purpose of the function is to mutate the object and continue doing so recursively. Right?